00XX-common-ledger-protocol/00XX-common-ledger-protocol.md

```diff
@@ -0,0 +1,273 @@
+
+In lots of scenarios, we don't have an underlying ledger that's fast enough to
+do every ILP payment on-ledger. There are also many situations when two parties
+trust one another up to an amount, and don't need any third-party ledger to
+book-keep for them. Currently, this need is filled by [ILP Plugin Virtual],
+which uses [Plugin RPC] as its ledger protocol. JSON messages are passed back
+and forth to perform transfers and send quote requests. Custom JSON messages
+can be added to extend the Plugin RPC messages, but they aren't well defined.
+
+Because the use case for a bilateral ledger protocols is so ubiquitous, the CLP
+has been invented to make the protocol extensible, efficient, and friendly to
+re-implementation.
+
+### Scope
+
+The CLP, despite its name, is just another ledger protocol. It manages
+conditional transfers, messaging requests, and allows [side protocols] for
+extensibility. You can use ILP without using CLP. CLP is intended to be a
+well-suited solution so that a new bilateral ledger protocol doesn't need to
+exist for every new use case.
+
+This document describes the flow and messages that CLP uses. It does not cover
+establishment of CLP channels, nor does it cover any side protocols.  Side
+protocols include functionality like ledger metadata, balance, automated
+settlement, and dispute resolution. 
+
+The CLP Packet format is described exactly in the [CLP ASN.1 spec].
+
+## Terminology
+
+- **CLP** is the common ledger protocol, as described by this packet and the
+  ASN.1 spec.
+
+- A **Side Protocol** is a protocol which isn't defined by CLP and is carried
+  in the side protocol data (see below).
+
+- A **CLP Connection** is a websocket connection over which CLP Packets are
+  sent.
+
+- **CLP Packets** are the messages defined in the CLP ASN.1 spec.
+
+- **Peers** are the parties on a CLP connection. Your peer is the party on the
+  other side of the CLP connection.
+
+- The **Bilateral Ledger** is the ledger which a CLP connection controls.  The
+  bilateral ledger is a persistent log of CLP packets, which can be used to
+deduce the current balance between two peers and the state of all transfers
+between them.
+
+- **Authoritative State** means that the entire Bilateral Ledger is being
+  tracked by a party. Both peers on a CLP connection can keep authoritative
+state, which means they can get into dispute if they disagree on the state of a
+transfer. This usually happens when network latency causes the peers to
+disagree about expiries. If one party keeps authoritative state, the other
+party must trust them not to tamper with it.
+
+## Overview
+
+CLP is broken up into 4 different RPC requests, which can get 2 different
+responses. Every CLP packet follows a common structure:
+
+```
+  +---------------+
+1 | Type (1)      |
+  +---------------+
+1 | Request ID    |
+2 | (2)           |
+3 |               |
+4 |               |
+  +---------------+
+1 | Length Prefix |
+2 | (3)           |
```

Comment:
I think this length prefix isn't there, because packet-specific data is either empty or a `SEQUENCE`, and sequences don't have lenght prefixes. See https://github.com/interledgerjs/clp-packet/blob/mj-testnet-1/test/indexSpec.js#L58

Reply:
the fifth byte in those test cases is this length prefix

Reply:
Oh wow, I think you mean the sixth byte, i.e. `buffer[5]`, and that indeed looks like one :) Also, https://github.com/interledger/rfcs/blob/master/asn1/CommonLedgerProtocol.asn#L118 states that packet.data should be length-prefixed. But from how I understand ASN.1/OER, that's not correct, is it? There is no reason to add a length-prefix to packet.data, because it's a sequence and you know which fields to expect.

Reply:
So I think that's a bug, then.

Reply:
It's not a bug, ASN.1 Open Types get a length prefix in OER (see section 30 of https://www.itu.int/rec/T-REC-X.696-201408-S/en)